File: propose-votes/manage-proposals.py
#!/usr/bin/python3

"""
Script language: Python3

Talks to:
- Vega node (REST)
- Vega wallet (REST)

Apps/Libraries:
- REST: requests (https://pypi.org/project/requests/)
"""

# Note: this file uses smart-tags in comments to section parts of the code to
# show them as snippets in our documentation. They are not necessary to be
# included when creating your own custom code.
#
# Example of smart-tags:
#  __something:
# some code here
# :something__

import json
import requests
import time
import os
import helpers

node_url_rest = os.getenv("NODE_URL_REST")
if not helpers.check_url(node_url_rest):
    print("Error: Invalid or missing NODE_URL_REST environment variable.")
    exit(1)

wallet_server_url = os.getenv("WALLETSERVER_URL")
if not helpers.check_url(wallet_server_url):
    print("Error: Invalid or missing WALLETSERVER_URL environment variable.")
    exit(1)

wallet_name = os.getenv("WALLET_NAME")
if not helpers.check_var(wallet_name):
    print("Error: Invalid or missing WALLET_NAME environment variable.")
    exit(1)

wallet_passphrase = os.getenv("WALLET_PASSPHRASE")
if not helpers.check_var(wallet_passphrase):
    print("Error: Invalid or missing WALLET_PASSPHRASE environment variable.")
    exit(1)

# Help guide users against including api version suffix on url
wallet_server_url = helpers.check_wallet_url(wallet_server_url)

#####################################################################################
#                           W A L L E T   S E R V I C E                             #
#####################################################################################

print(f"Logging into wallet: {wallet_name}")

# __login_wallet:
# Log in to an existing wallet
req = {"wallet": wallet_name, "passphrase": wallet_passphrase}
response = requests.post(f"{wallet_server_url}/api/v1/auth/token", json=req)
helpers.check_response(response)
token = response.json()["token"]
# :login_wallet__

assert token != ""
print("Logged in to wallet successfully")

# __get_pubkey:
# List key pairs and select public key to use
headers = {"Authorization": f"Bearer {token}"}
response = requests.get(f"{wallet_server_url}/api/v1/keys", headers=headers)
helpers.check_response(response)
keys = response.json()["keys"]
pubkey = keys[0]["pub"]
# :get_pubkey__

assert pubkey != ""
print("Selected pubkey for signing")

#####################################################################################
#                            L I S T   P R O P O S A L S                            #
#####################################################################################

# There are two types of REST request for proposals on Vega:
# 1 - MARKET proposals (/governance/market/proposals)
# 2 - ASSET proposals (/governance/asset/proposals)
# Note: In the future users will be able to call an endpoint to retrieve ALL proposals.

# __get_proposals:
# Request a list of proposals on a Vega network
response = requests.get(node_url_rest + "/governance/market/proposals")
helpers.check_response(response)
proposals = response.json()
print("Proposals:\n{}".format(json.dumps(proposals, indent=2, sort_keys=True)))
# :get_proposals__

proposalID = proposals["data"][0]["proposal"]["id"]
assert proposalID != ""
print(f"Proposal found: {proposalID}")

#####################################################################################
#                         P R O P O S A L   D E T A I L S                           #
#####################################################################################

# __get_proposal_detail:
# Request results of a specific proposal on a Vega network
response = requests.get(node_url_rest + "/governance/proposal/" + proposalID)
helpers.check_response(response)
response_json = response.json()
print("Proposal:\n{}".format(json.dumps(response_json, indent=2, sort_keys=True)))
# :get_proposal_detail__

#####################################################################################
#                          P A R T Y   P R O P O S A L S                            #
#####################################################################################

# __get_proposals_by_party:
# Request results of a specific proposal on a Vega network
response = requests.get(node_url_rest + "/parties/" + pubkey + "/proposals")
helpers.check_response(response)
response_json = response.json()
print("Party proposals:\n{}".format(json.dumps(response_json, indent=2, sort_keys=True)))
# :get_proposals_by_party__

File: propose-votes/manage-proposals.go
package main

import (
	"encoding/json"
	"fmt"
	"io/ioutil"
	"net/http"
	"os"

	"github.com/vegaprotocol/api-clients/go/generated/code.vegaprotocol.io/vega/proto/api"
	"golang.org/x/net/context"
	"google.golang.org/grpc"
	"code.vegaprotocol.io/go-wallet/wallet"
)

func main() {
	nodeURLGrpc := os.Getenv("NODE_URL_GRPC")
	if len(nodeURLGrpc) == 0 {
		panic("NODE_URL_GRPC is null or empty")
	}
	walletserverURL := os.Getenv("WALLETSERVER_URL")
	if len(walletserverURL) == 0 {
		panic("WALLETSERVER_URL is null or empty")
	}
	walletName := os.Getenv("WALLET_NAME")
	if len(walletName) == 0 {
		panic("WALLET_NAME is null or empty")
	}
	walletPassphrase := os.Getenv("WALLET_PASSPHRASE")
	if len(walletPassphrase) == 0 {
		panic("WALLET_PASSPHRASE is null or empty")
	}

	walletserverURL = CheckWalletUrl(walletserverURL)

	walletConfig := WalletConfig{
		URL:        walletserverURL,
		Name:       walletName,
		Passphrase: walletPassphrase,
	}

	conn, err := grpc.Dial(nodeURLGrpc, grpc.WithInsecure())
	if err != nil {
		panic(err)
	}
	defer conn.Close()

	dataClient := api.NewTradingDataServiceClient(conn)

	var token wallet.TokenResponse
	body, err := LoginWallet(walletConfig)
	if err != nil {
		panic(err)
	}
	json.Unmarshal([]byte(body), &token)

	fmt.Println(token.Token)

	// List existing keypairs
	url := walletserverURL + "/api/v1/keys"
	req, err := http.NewRequest(http.MethodGet, url, nil)
	req.Header.Set("Authorization", "Bearer "+token.Token)

	client := &http.Client{}
	resp, err := client.Do(req)
	if err != nil {
		panic(err)
	}
	defer resp.Body.Close()

	body, err = ioutil.ReadAll(resp.Body)
	if err != nil {
		panic(err)
	}
	fmt.Println("response Body:", string(body))
	var keypair wallet.KeysResponse
	json.Unmarshal([]byte(body), &keypair)

	if len(keypair.Keys) == 0 {
		panic("No keys!")
	}

	pubkey := keypair.Keys[0].Pub
	fmt.Println("pubkey: ", pubkey)

	// List proposals
	// __get_proposals:
	// Request a list of proposals on a Vega network

	proposalsRequest := api.GetProposalsRequest{}
	proposalsRequest = api.GetProposalsRequest{}
	proposalResponse, err := dataClient.GetProposals(context.Background(), &proposalsRequest)
	fmt.Printf("Proposals: %+v\n", proposalResponse)
	// :get_proposals__

	if len(proposalResponse.Data) == 0 {
		fmt.Println("No open proposal, exit!")
		return
	}

	proposalID := proposalResponse.Data[0].Proposal.Id

	// Get proposal details
	// __get_proposal_detail:
	// Request results of a specific proposal on a Vega network
	proposalByIDRequest := api.GetProposalByIDRequest{ProposalId: proposalID}
	proposalByIDResponse, err := dataClient.GetProposalByID(context.Background(), &proposalByIDRequest)

	fmt.Printf("ProposalById: %+v\n", proposalByIDResponse)
	// :get_proposal_detail__

	// Party proposals
	// __get_proposals_by_party:
	// Request results of a specific proposal on a Vega network
	proposalsByPartyIDRequest := api.GetProposalsByPartyRequest{PartyId: pubkey}
	proposalsByPartyIDResponse, err := dataClient.GetProposalsByParty(context.Background(), &proposalsByPartyIDRequest)

	fmt.Printf("ProposalByPartyId: %+v\n", proposalsByPartyIDResponse)
	// :get_proposals_by_party__

}
